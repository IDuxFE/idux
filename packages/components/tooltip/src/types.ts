import type { DefineComponent } from 'vue'
import type { OverlayPlacement, OverlayTrigger } from '@idux/cdk/overlay'

import { PropTypes } from '@idux/cdk/utils'
import { overlayPlacementDef, overlayTriggerDef } from '@idux/cdk/overlay'

export interface TooltipProps {
  /**
   * title of popover
   */
  title?: string
  /**
   * alignment between trigger and popover
   */
  placement?: OverlayPlacement
  /**
   * visibility of popover
   */
  visible?: boolean
  /**
   * the way to toggle popover
   */
  trigger?: OverlayTrigger
  showDelay?: number
  hideDelay?: number
  /**
   * whether destroy when popover is hidden
   */
  destroyOnHide?: boolean
  /**
   * whether re-adjust when popover is be covered
   */
  autoAdjust?: boolean
}

export const tooltipPropsDef = {
  title: PropTypes.string,
  placement: overlayPlacementDef,
  visible: PropTypes.bool.def(false),
  trigger: overlayTriggerDef,
  hideDelay: PropTypes.number,
  showDelay: PropTypes.number,
  destroyOnHide: PropTypes.bool,
  autoAdjust: PropTypes.bool,
}

export type TooltipComponent = InstanceType<DefineComponent<TooltipProps>>
