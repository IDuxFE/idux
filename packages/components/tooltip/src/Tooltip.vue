<template>
  <ix-overlay
    v-model:visible="visibility"
    clsPrefix="ix-tooltip"
    allow-enter
    v-bind="config"
    scroll-strategy="none"
    :offset="offset"
  >
    <template #trigger>
      <slot />
    </template>
    <template #overlay>
      <slot name="title">{{ title }}</slot>
    </template>
  </ix-overlay>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import { IxOverlay } from '@idux/components/overlay'

import { tooltipPropsDef } from './types'
import { useConfig, useOffset, useVisibility } from './utils'

export default defineComponent({
  name: 'IxTooltip',
  components: { IxOverlay },
  props: tooltipPropsDef,
  emits: ['update:visible'],
  setup() {
    const config = useConfig()
    const visibility = useVisibility()
    const offset = useOffset(config)

    return { visibility, config, offset }
  },
})
</script>
