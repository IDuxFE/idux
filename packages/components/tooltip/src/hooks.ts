import type { ComputedRef, WritableComputedRef } from 'vue'
import type { TooltipConfig } from '@idux/components/config'
import type { TooltipProps } from './types'

import { computed, getCurrentInstance } from 'vue'
import { useGlobalConfig } from '@idux/components/config'
import { callEmit } from '@idux/cdk/utils'

export function useConfig(): ComputedRef<TooltipConfig> {
  const config = useGlobalConfig('tooltip')
  const props = getCurrentInstance()!.props as TooltipProps

  return computed(() => {
    return {
      placement: props.placement ?? config.placement,
      trigger: props.trigger ?? config.trigger,
      showDelay: props.showDelay ?? config.showDelay,
      hideDelay: props.hideDelay ?? config.hideDelay,
      destroyOnHide: props.destroyOnHide ?? config.destroyOnHide,
      autoAdjust: props.autoAdjust ?? config.autoAdjust,
    }
  })
}

export function useVisibility(): WritableComputedRef<boolean> {
  const props = getCurrentInstance()!.props as TooltipProps

  return computed({
    get() {
      return !!props.visible
    },
    set(value) {
      callEmit(props['onUpdate:visible'], value)
    },
  })
}
