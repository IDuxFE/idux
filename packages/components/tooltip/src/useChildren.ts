import type { Slots, VNode } from 'vue'

import { getFirstValidNode, getSlotNodes } from '@idux/cdk/utils'

export function useChildren(slots: Slots, key?: string): VNode | undefined {
  return getFirstValidNode(getSlotNodes(slots, key), 1)
}
