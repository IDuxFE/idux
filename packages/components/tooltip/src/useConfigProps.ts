/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { TooltipProps } from './types'
import type { TooltipConfig } from '@idux/components/config'
import type { ComputedRef } from 'vue'

import { computed } from 'vue'

export function useConfigProps(
  props: TooltipProps,
  config: TooltipConfig,
  setVisibility: (visible: boolean) => void,
): ComputedRef<TooltipConfig & { clickOutside: boolean }> {
  return computed(() => {
    const trigger = props.trigger ?? config.trigger
    return {
      autoAdjust: props.autoAdjust ?? config.autoAdjust,
      clickOutside: trigger === 'click' || trigger === 'contextmenu',
      delay: props.delay ?? config.delay,
      destroyOnHide: props.destroyOnHide ?? config.destroyOnHide,
      placement: props.placement ?? config.placement,
      target: props.target ?? config.target,
      trigger: trigger,
      ['onUpdate:visible']: setVisibility,
    }
  })
}
