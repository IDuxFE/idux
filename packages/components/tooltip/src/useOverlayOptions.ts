import type { OverlayOptions } from '@idux/cdk/overlay'
import type { GlobalConfigKey } from '@idux/components/config'
import type { TooltipProps } from './types'

import { getCurrentInstance } from 'vue'
import { useGlobalConfig } from '@idux/components/config'

export function useOverlayOptions(comp: Extract<'tooltip', GlobalConfigKey> = 'tooltip'): OverlayOptions {
  const props: TooltipProps = getCurrentInstance()!.props
  const config = useGlobalConfig(comp)

  return {
    visible: props.visible,
    scrollStrategy: 'reposition',
    showArrow: true,
    placement: props.placement ?? config.placement,
    trigger: props.trigger ?? config.trigger,
    allowEnter: true,
    offset: [0, 5],
    hideDelay: props.hideDelay ?? config.hideDelay,
    showDelay: props.showDelay ?? config.showDelay,
  }
}
