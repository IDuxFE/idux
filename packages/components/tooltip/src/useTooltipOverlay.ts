/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { TooltipProps } from './types'
import type { ɵOverlayInstance, ɵOverlayProps } from '@idux/components/_private/overlay'
import type { TooltipConfig } from '@idux/components/config'
import type { ComputedRef, Ref } from 'vue'

import { computed, ref } from 'vue'

import { useControlledProp } from '@idux/cdk/utils'

const defaultOffset: [number, number] = [0, 12]

export interface TooltipOverlayContext {
  overlayRef: Ref<ɵOverlayInstance | undefined>
  updatePopper: () => void | undefined
  visible: ComputedRef<boolean>
  setVisible: (value: boolean) => void
  overlayProps: ComputedRef<ɵOverlayProps>
}

export function useTooltipOverlay(
  props: TooltipProps,
  config: TooltipConfig,
  mergedPrefixCls: ComputedRef<string>,
): TooltipOverlayContext {
  const overlayRef = ref<ɵOverlayInstance>()
  const updatePopper = () => overlayRef.value?.updatePopper()

  const [visible, setVisible] = useControlledProp(props, 'visible', false)
  const overlayProps = computed(() => {
    const trigger = props.trigger ?? config.trigger
    return {
      visible: visible.value,
      ['onUpdate:visible']: setVisible,
      autoAdjust: props.autoAdjust ?? config.autoAdjust,
      clickOutside: trigger === 'click' || trigger === 'contextmenu',
      delay: props.delay ?? config.delay,
      destroyOnHide: props.destroyOnHide ?? config.destroyOnHide,
      offset: defaultOffset,
      showArrow: true,
      placement: props.placement ?? config.placement,
      target: props.target ?? config.target ?? `${mergedPrefixCls.value}-container`,
      trigger: trigger,
      zIndex: props.zIndex,
    }
  })
  return { overlayRef, updatePopper, visible, setVisible, overlayProps }
}
