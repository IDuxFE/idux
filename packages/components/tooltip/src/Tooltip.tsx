/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { TooltipProps } from './types'
import type { Slots } from 'vue'

import { computed, defineComponent } from 'vue'

import { ɵOverlay } from '@idux/components/_private/overlay'
import { useGlobalConfig } from '@idux/components/config'

import { tooltipProps } from './types'
import { useTooltipOverlay } from './useTooltipOverlay'

export default defineComponent({
  name: 'IxTooltip',
  props: tooltipProps,
  setup(props, { slots, expose }) {
    const common = useGlobalConfig('common')
    const config = useGlobalConfig('tooltip')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-tooltip`)
    const { overlayRef, updatePopper, overlayProps } = useTooltipOverlay(props, config, common, mergedPrefixCls)
    expose({ updatePopper })

    return () => {
      const prefixCls = mergedPrefixCls.value

      return (
        <ɵOverlay
          ref={overlayRef}
          v-slots={{ default: slots.default, content: () => renderContent(props, slots, prefixCls) }}
          class={prefixCls}
          transitionName={`${common.prefixCls}-fade-fast`}
          {...overlayProps.value}
        />
      )
    }
  },
})

const renderContent = (props: TooltipProps, slots: Slots, prefixCls: string) => {
  if (!(slots.title || props.title)) {
    return null
  }
  return <div class={`${prefixCls}-wrapper`}>{slots.title?.() ?? props.title}</div>
}
