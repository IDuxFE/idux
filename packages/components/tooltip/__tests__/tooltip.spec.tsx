import { mount, MountingOptions, VueWrapper } from '@vue/test-utils'
import { renderWork, wait } from '@tests'
import IxTooltip from '../src/Tooltip.vue'
import { TooltipProps } from '../src/types'

describe('Tooltip.vue', () => {
  let TooltipMount: (options?: MountingOptions<Partial<TooltipProps>>) => VueWrapper<InstanceType<typeof IxTooltip>>

  beforeEach(() => {
    TooltipMount = options => mount(IxTooltip, { ...options })
  })

  renderWork(IxTooltip)

  test('placement', async () => {
    jest.spyOn(console, 'warn').mockImplementation(() => {})
    const tooltipWrapper = TooltipMount({
      slots: {
        default: <span id="trigger">Hello world</span>,
      },
      props: {
        title: 'prompt text',
      },
    })
    await tooltipWrapper.get('#trigger').trigger('mouseenter')
    await wait(100)
    expect(document.querySelector('.ix-tooltip')!.getAttribute('placement')).toEqual('top')

    await tooltipWrapper.setProps({ placement: 'bottom' })
    expect(document.querySelector('.ix-tooltip')!.getAttribute('placement')).toEqual('bottom')

    await tooltipWrapper.setProps({ placement: 'left' })
    expect(document.querySelector('.ix-tooltip')!.getAttribute('placement')).toEqual('left')

    await tooltipWrapper.setProps({ placement: 'right' })
    expect(document.querySelector('.ix-tooltip')!.getAttribute('placement')).toEqual('right')
  })
})
