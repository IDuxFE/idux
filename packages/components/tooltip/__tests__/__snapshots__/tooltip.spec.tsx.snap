// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip.vue render work 1`] = `"<!---->"`;
